File: C#/Leetcode/Array/SetMatrixZeroes.cs
namespace LeetcodeSolutions.Array
{
    // Leetcode 73
    // Submission Detail: https://leetcode.com/submissions/detail/151101702/
    public class SetMatrixZeroes
    {
        //[                   [
        //  [1,1,1],            [1,0,1],
        //  [1,0,1],    =>      [0,0,0],
        //  [1,1,1]             [1,0,1]
        //]                    ]

        // Tx = O(mn)
        // Sx = O(1)
        // Idea: 1. Use firstrow and firstcolumn flags to check if the values need to be set to 0.
        //          if (i,j) is 0, set (i,0) and (0,j) to 0 in first iteration. Then,
        //          set the (i,j) to 0 if (i,0) and (0,j) was previously set to 0.
        public void SetZeroes(int[,] matrix)
        {
            int m = matrix.GetLength(0);
            int n = matrix.GetLength(1);

            bool firstRow = false;
            bool firstColumn = false;

            for (int i = 0; i < m; i++)
            {
                for (int j = 0; j < n; j++)
                {
                    if (matrix[i, j] == 0)
                    {
                        if (i == 0) firstRow = true;
                        if (j == 0) firstColumn = true;
                        if (i != 0 && j != 0)
                        {
                            matrix[i, 0] = 0;
                            matrix[0, j] = 0;
                        }
                    }
                }
            }

            for (int i = 1; i < m; i++)
                for (int j = 1; j < n; j++)
                    if (matrix[0, j] == 0 || matrix[i, 0] == 0)
                        matrix[i, j] = 0;

            if (firstRow)
                for (int j = 0; j < n; j++)
                    matrix[0, j] = 0;

            if (firstColumn)
                for (int i = 0; i < m; i++)
                    matrix[i, 0] = 0;
        }
    }
}

File: Java/Matrices/SetMatrixZeroes.java
package Matrices;

public class SetMatrixZeroes {
  // Leetcode 73: https://leetcode.com/problems/set-matrix-zeroes/description/
  // Submission Detail: https://leetcode.com/problems/set-matrix-zeroes/submissions/1659443321

  //  Input: matrix = [[1,1,1],[1,0,1],[1,1,1]]
  //  Output: [[1,0,1],[0,0,0],[1,0,1]]

  //  Input: matrix = [[0,1,2,0],[3,4,5,2],[1,3,1,5]]
  //  Output: [[0,0,0,0],[0,4,5,0],[0,3,1,0]]

  public void setZeroes(int[][] matrix) {
    boolean firstRow = false, firstColumn = false;

    // Step 1: First Pass - iterate and track what cols, rows needs to be set to 0.
    for(int row=0; row<matrix.length; row++) {
      for(int col=0; col<matrix[0].length; col++) {
        if(matrix[row][col] == 0) {
          if (row == 0)
            firstRow = true;

          if(col == 0)
            firstColumn = true;

          if(row != 0 && col != 0) {
            matrix[0][col] = 0;
            matrix[row][0] = 0;
          }
        }
      }
    }

    // Step 2: Use the tracked values to set the cell value to 0
    for(int row=0; row< matrix.length; row++) {
      for(int col = 0; col < matrix[0].length; col++) {
        if(row != 0 && col != 0 && (matrix[row][0] == 0 || matrix[0][col] == 0))
          matrix[row][col] = 0;
      }
    }

    // Step 3: Use the tracked values to set the firstCol value to 0
    if(firstColumn) {
      for(int row=0; row< matrix.length; row++)
        matrix[row][0] = 0;
    }

    // Step 4: Use the tracked values to set the firstRow value to 0
    if(firstRow){
      for(int col = 0; col < matrix[0].length; col++)
        matrix[0][col] = 0;
    }
  }
}
